### PR TITLE
Added parent TokenSale contract

### PR DESCRIPTION
**Pull request by alibros**
Parent Contract that will be extended by `PreSale` and `General Sale`

### CODE REVIEW

File: contracts/TokenSale.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    address public etherEscrowAddress;
+    address public reserveAddress;
+    address public founderAddress;
+
+    uint256 constant public CALLER_EXCHANGE_RATE = 2000;
+    uint256 constant public FOUNDER_EXCHANGE_RATE = 1000;
+    uint256 constant public RESERVE_EXCHANGE_RATE = 2000;
+    uint256 constant public MAX_GAS_PRICE = 50000000000;
+
+    bool public paused;
```
Comment (israelcolomer) [inline contracts/TokenSale.sol]: bool public closed; needed too

File: contracts/TokenSale.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    uint256 public finalizedBlock;
```
Comment (israelcolomer) [inline contracts/TokenSale.sol]: Not sure we'll need these 2, but we can leave them here for the time being.

File: contracts/TokenSale.sol [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    uint256 public finalizedTime;
+
+    event NewSale(address indexed _caller, uint256 etherAmount, uint256 tokensGenerated);
```
Comment (israelcolomer) [inline contracts/TokenSale.sol]: I would rename to Sale or TokenSale?

File: contracts/TokenSale.sol [hunk 4]
```diff
@@ -0,0 +1,126 @@
+pragma solidity ^0.4.11;
+
+import "./lib/SafeMath.sol";
+import "./lib/Owned.sol";
+import "./SHP.sol";
+import "./SCD.sol";
+import "./AffiliateUtility.sol";
+
+
+contract TokenSale is Owned {
```
Comment (israelcolomer) [inline contracts/TokenSale.sol]: Add a validTransaction modifier?

File: contracts/TokenSale.sol [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    using SafeMath for uint256;
+    
+    SHP public shp;
+    AffiliateUtility public affiliateUtility;
```
Comment (israelcolomer) [inline contracts/TokenSale.sol]: How do we set this contract?

File: contracts/TokenSale.sol (hunk 4 above)
Comment (alibros) [reply contracts/TokenSale.sol]: 👍

File: contracts/TokenSale.sol (hunk 5 above)
Comment (alibros) [reply contracts/TokenSale.sol]: in the constructor, pass the address of an already deployed one. similar with SHP

File: contracts/TokenSale.sol [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    /// @notice Resumes the contribution
+    function resumeContribution() public payable onlyOwner {
+        // TODO: enable
+        // ContributionStateChanged(msg.sender, ContributionState.Resumed);
+        isSaleOpen = true;
+    }
+
+    function bytesToAddress (bytes b) internal constant returns (address) {
```
Comment (lewisbarber) [inline contracts/TokenSale.sol]: Should it be constant?

File: contracts/TokenSale.sol [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            assembly {
+                size := extcodesize(_caller)
+            }
+            return (size > 0);
+        }
+    }
+
+    /// @notice Pauses the contribution if there is any issue
+    function pauseContribution() public payable onlyOwner {
+        // TODO: enable
```
Comment (lewisbarber) [inline contracts/TokenSale.sol]: TODO?

File: contracts/TokenSale.sol [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        // ContributionStateChanged(msg.sender, ContributionState.Paused);
+        isSaleOpen = false;
+    }
+
+    /// @notice Resumes the contribution
+    function resumeContribution() public payable onlyOwner {
+        // TODO: enable
```
Comment (lewisbarber) [inline contracts/TokenSale.sol]: TODO?

**Pull request merged**